SCP-2232: Chain index types

j-mueller opened the pull request:
Starting to separate the chain index from the PAB. See https://hackmd.io/ji8Hs_88TlqMOK0VeX9Row for the original specification.

This only defines the types and query interface, implementation still TBD.

<!--
Here are some checklists you may like to use. Use your judgement.

This is just a checklist, all the normative suggestions are covered in more detail in CONTRIBUTING.
-->
Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

sjoerdvisscher commented on plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs:
```diff
@@ -0,0 +1,49 @@
+        let newlySpent = _usUnmatchedSpentOutputs r `Set.difference` _usUnmatchedSpentOutputs l in
+        UtxoState
+            { _usUnspentOutputs = (_usUnspentOutputs l <> _usUnspentOutputs r) `Set.difference` newlySpent
+            , _usUnmatchedSpentOutputs = _usUnmatchedSpentOutputs l <> (_usUnmatchedSpentOutputs r `Set.difference` newlySpent)
```
I drew a venn diagram and this seems to be the same as just `_usUnmatchedSpentOutputs l`. `(_usUnmatchedSpentOutputs r `Set.difference` newlySpent)` is the part of `r` that is already in `l`.

I would have thought this would actually be commutative. (`<>` as a whole isn't because of the slot and blockId fields)

```
spent = (spent l - unspent r) <> (spent r - unspent l)
unspent = (unspent l - spent r) <> (unspent r - spent l)
```

j-mueller replied:
Fine I'll add a test ;) 

I don't think it can be commutative - inputs must be produced *before* they can be spent, so every TxOutRef must eventually appear in a `_usUnspentOutputs` field before it appears in an `_usUnmatchedSpentOutputs`.

sjoerdvisscher replied:
That just means there can't be any overlap between `spent l` and `unspent r`? So we can simplify to

```
spent = spent l <> (spent r - unspent l)
unspent = (unspent l - spent r) <> unspent r
```

and then it is no longer commutative using the assumption.

sjoerdvisscher commented on plutus-chain-index/src/Plutus/ChainIndex/Query.hs:
```diff
@@ -0,0 +1,52 @@
+    -- | Get the monetary policy from an MPS hash (if available)
+    MonetaryPolicyFromHash :: MonetaryPolicyHash -> ChainIndexQueryEffect (Maybe MonetaryPolicy)
+
+    -- | Get the TxOut from a TxOutRef if available
```
```suggestion
    -- | Get the TxOut from a TxOutRef (if available)
```

For consistency :)

sjoerdvisscher commented on plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs:
```diff
@@ -0,0 +1,49 @@
+    UtxoState
+        { _usUnspentOutputs        :: Set TxOutRef -- ^ Outputs that have not been spent yet
+        , _usUnmatchedSpentOutputs :: Set TxOutRef -- ^ Outputs that we have seen a spending transaction for, but not a producing transaction
+        , _usSlot                  :: Last Slot -- ^ Last slot that we have seen
+        , _usBlockId               :: Last BlockId -- ^ Last block ID
```
Maybe `_usTip :: Last (Slot, Maybe BlockId)` is more correct? Not sure.

j-mueller replied:
Yes that's better, thanks.

j-mueller commented:
@sjoerdvisscher I fixed the definitions and added a property test. This wasn't straightforward because I had to write the generators to produce the right data. But it seems to be good now.

sjoerdvisscher commented:
Some of these generators needed to be written anyway, right?
I really like the code overall. 💯

j-mueller commented:
@michaelpj this is just hanging on a VM test, could please give it a little nudge? Thank you!
